packages/@ember/application/lib/application.js

```diff
@@ -483,13 +483,22 @@ const Application = Engine.extend({
     @method waitForDOMReady
   */
   waitForDOMReady() {
-    if (!this.$ || this.$.isReady) {
+    if (!hasDOM || this.documentReadyState() !== 'loading') {
```

Comment:
@chancancode Shouldn't `!hasDom` guard us against accessing `document` in the fastboot environment in addition to `if (this.autoboot) { this.waitForDOMReady() }`?

```js
// packages/@ember/-internals/browser-environment/lib/has-dom.ts

export default typeof self === 'object' &&
  self !== null &&
  (self as any).Object === Object &&
  typeof Window !== 'undefined' &&
  self.constructor === Window &&
  typeof document === 'object' &&
  document !== null &&
  self.document === document &&
  typeof location === 'object' &&
  location !== null &&
  self.location === location &&
  typeof history === 'object' &&
  history !== null &&
  self.history === history &&
  typeof navigator === 'object' &&
  navigator !== null &&
  self.navigator === navigator &&
  typeof navigator.userAgent === 'string';
```

Reply:
@rreckonerr good point! so it's already guarded 👍